### json 불러와서 저장하기

In [3]:
import numpy as np
import pandas as pd
import json
import jpype
import glob
from random import shuffle
from bs4 import BeautifulSoup as bs

data_path = "data"
#data_path = "data(test)"

file_list = glob.glob("%s/*.json" % data_path)
shuffle(file_list)

json_train=[]
for json_file_name in file_list:
    json_file = json.loads(open(json_file_name).read())
    json_train += json_file["articles"]

### json에서 text 형태소 추출

In [4]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    text = bs(article["text"], "html.parser").text
    
    if cnt % 10000 == 0:
        print(cnt)
        
    text_pos = ["%s_%s" % (first, second) for first, second in mecab.pos(text)]

    labeled_train.append({
            "istroll": article["is_troll"],
            "text_pos": text_pos,
            "text_pos_sentences" : " ".join(text_pos),
            "pk": article["pk"]
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train = labeled_train.set_index('pk')

labeled_train

0
10000
20000
30000
40000

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://news.sportsseoul.com/read/entertain/1295201.htm<div><span><br><br>\xed\x95\x98\xec\x95\x84...</span></div>'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://media.daum.net/society/others/newsview?rMode=list&cSortKey=depth&allComment=T&newsid=20140103112805017'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"


50000
60000

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://www.4shared.com/audio/T_z57p79/Pendulum_-_Blood_Sugar_wwwshuf.htm\r\n\r\n\xe3\x85\x8b\xe3\x85\x8b\xe3\x85\x8b\xe3\x85\x8b'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://www.ohmynews.com/NWS_Web/View/at_pg.aspx?CNTN_CD=A0001947743'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beaut


70000

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:198: UserWarning: "b'..'" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.
  '"%s" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://blog.naver.com/kakashi08/110086751501'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserW


80000

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.bgmstore.net/view/049Ju'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.news.naver.com/read.nhn?sid1=100&oid=055&aid=0000269444'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind 


90000
100000
110000
120000
130000
140000
150000

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://imgur.com/rBtm1LI.jpg'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.ilwar.com/poli/17062'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beau


160000

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://ilwar.com/poli/53209'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://1.hidemyass.com/ip-9/encoded/Oi8vd3d3LnJvZG9uZy5yZXAua3Ava28v'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind t


170000


/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.khan.co.kr/view.html?category=2&med_id=khan&artid=201401071905352&code=361102'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)


istroll                                           text_pos  \
pk                                                                 
86954    True  [내_NP, 란_VCP+ETM, 음모_NNG, 사건_NNG, 17_SN, 일_NNB...   
86956   False  [뭔가_NP+VCP+EC, 벤쿠버_NNP, 때_NNG, 의_JKG, 느낌_NNG, ...   
86957   False  [그건_NP+JX, 전두환_NNP, 이_JKS, 일으킨_VV+ETM, 폭동_NNG,...   
86958   False  [아_IC, 웃_VV+EP, 곀ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ_UNKNOWN, 인조인간_N...   
86959   False  [정치_NNG, 와_JC, 경제_NNG, 와_JC, 사회_NNG, 는_JX, 하나_...   
86960   False  [이_MM, 사람_NNG, 이_JKS, 참여_NNG, 정부_NNG, 때_NNG, 까...   
86962   False  [그래봐야_VA+EC+VX+EC,  _SY, 넌_NP+JX, 이미_MAG, 내_NP...   
86963   False  [m_SL, ._SY, news_SL, ._SY, naver_SL, ._SY, co...   
86964   False  [정말_MAG, 조_NNG, 같_VA, 은_ETM, 나라_NNG, 네요_VCP+EC...   
86965   False  [글쓰기_NNG, 공지_NNG, 사항_NNG, 을_JKO, 보_VV, 니까_EC, ...   
86966   False  [홍대_NNG, 역_NNG, 2_SN, 번_NNBC, 출구_NNG, 에서_JKB, ...   
86967   False  [자극_NNG, 적_XSN, 인_VCP+ETM, 기사_NNG, 네_XSN, 낚시_N...   
86968   False               [나농_NNG, 은_JX, 베트남_NNP, ._SF, .._SY]   
86969   False  [http_SL, :_SC, //_SY, www_SL, ._SY, ilwar_SL,...   
86970   False  [ㅠㅠ_UNKNOWN, 내_NP, 가_JKS, 그리_MAG, 잘못_MAG, 한_XS...   
86971   False  [집_NNG, 에_JKB, 도_JX, 가깝_VA, 고_EC, 경쟁_NNG, 률_XS...   
86973   False                                                 []   
86975   False  [듣보잡_NNG, 레볼루션_NNP, 뉴스_NNG, 조선일보_NNP, 에_JKB, 직...   
86976   False  [웬지_MAG, 무섭_VA, 게_EC, 느껴_VV+EC, 지_VX, 는_ETM, 건...   
86978   False  [뜨신_VA+ETM, 겨울_NNG, 코트_NNG, 사_VV+EC, 주_VX, 고_E...   
86979   False                              [반갑_VA, 습니다_EF, ._SF]   
86980   False  [중앙선관위_NNP, 의_JKG, “_SSO, 도장_NNG, 대란_NNG, ”_SS...   
86982   False  ["_SY, 쥐_NNG, 가_JKS, 싫_VA, 농_NNG, 닭_NNG, 이_JKS...   
86983   False  [내_NP, 가_JKS, 가장_MAG, 사랑_NNG, 하_XSV, 는_ETM, ,_...   
86984   False  [ _SY, 혹시나_MAG, 싶_VA, 어서_EC, 검색_NNG, 해_XSV+EC,...   
86985   False  [서점_NNG, 에서_JKB, 책_NNG, 보_VV, 고_EC, 있_VX, 는_ET...   
86986   False  [2013_SN, ._SY, 3_SN, ._SY, 17_SN, -_SY, 인천_NN...   
86987   False  [ _SY, 새로운_VA+ETM, 뉴스_NNG, 캐스트_NNG, 적용_NNG, 샷_...   
86989   False  [뭔_MM, 뜻_NNG, 인지_VCP+EC, 잘_MAG, 모르_VV, 겠_EC, 닭...   
86990   False  [이곳_NP, 일간_MAG, 워스트_NNP, 는_JX, 반일_NNG, 베_NNG, ...   
...       ...                                                ...   
42885   False           [뭐_IC, 그렇_VA, 닭_NNG, 잘_MAG, 지내_VV, 쟈_EC]   
42886   False  [사실_MAG, 본_VV+ETM, 농_NNG, 은_JX, 왜서_MAG, 방사능_NN...   
42887   False  [일간_NNG, 워스트_NNP, 회원_NNG, 수_NNG, 가_JKS, 많이_MAG...   
42888   False  [http_SL, :_SC, //_SY, www_SL, ._SY, wikitree_...   
42889   False  [벌레_NNG, 퇴치_NNG, 와_JC, 이_MM, 나라_NNG, 의_JKG, 모든...   
42890   False  [안녕_NNG, 하_XSV, 세요_EP+EF, ._SF, 가입_NNG, 인사_NNG...   
42891   False  [이번_NNG, 주말_NNG, 드디어_MAG, 변호인_NNG, 보_VV, 러_EC,...   
42892   False  [용용_IC, 귀농_NNG, 이_JKS, 닭_NNG, 와_VV+EC, 이리_MAG,...   
42893   False  [진짜_MAG, 어_IC, 이_NP, 가_JKS, 없_VA, 는_ETM, 글_NNG...   
42894   False  [ㅈ_NNG, ㅇㅇ_UNKNOWN, 눌러_VV+EC, 주_VX, 려니까_EC, 인사...   
42895   False  [소문_NNG, 듣_VV, 고_EC, 가입_NNG, 했_XSV+EP, 습니다_EF,...   
42896   False  [저_NP, 도_JX, 좋_VA, 아요_EF, 클릭_NNG, 하_XSV, 기_ETN...   
42897   False  [http_SL, :_SC, //_SY, mediawatch_SL, ._SY, kr...   
42898   False  [우리_NP, 청와대_NNP, 에서_JKB, 는_JX, 이_MM, 사이트_NNG, ...   
42899   False  [텃밭_NNG, 이나_JC, 뭐_IC, 뒷밭_NNG, 이런_MM, 걸로_NNB+JK...   
42900   False  [가입_NNG, 인사_NNG, 드린다_VV+EC, 농_NNG, ~~~_SY, 열심히...   
42902   False                   [귀농_NNG, 합_NNG, 니_VCP+EC, 닭_NNG]   
42903   False  [상식_NNG, 이_JKS, 통하_VV, 는_ETM, 사회_NNG, 가_JKS, 되...   
42904   False  [앞_NNG, 으로_JKB, 눈_NNG, 팅_MAG, 이_MM, 더_MAG, 즐거워...   
42905   False  [역사_NNG, 밭_NNG, 을_JKO, 잘_MAG, 키워_VV+EC, 보_VX, ...   
42906   False  [반갑_VA, 습니다_EF, ._SF, 사람_NNG, 소리_NNG, 가_JKS, 나...   
42907   False  [바깥_NNG, 에_JKB, 쓰레기_NNG, 버리_VV, 러_EC, 갔_VV+EP,...   
42908   False  [언제_MAG, 생겼_VV+EP, 나요_EF, ?_SF, 트위터_NNP, 보_VV,...   
42909   False  [의사_NNG, 가_JKS, 말_NNG, 하_XSV, 길_ETN+JKO, 볼_NNG...   
42910   False  [모두_MAG, 모두_MAG, 반갑_VA, 사오_EC, 오늘_MAG, 처음_MAG,...   
42912   False  [여기_NP, 는_JX, 오늘_MAG, 날씨_NNG, 가_JKS, 꿀꿀_MAG, 하.

In [5]:
used_model = ""

### bag of words

In [6]:
from sklearn.feature_extraction.text import CountVectorizer

def make_bag_of_words(labeled_train, max_features, col_name):
    global used_model
    used_model = "bow"
    
    vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                                stop_words = None, max_features=max_features)
    
    train_data_features = vectorizer.fit_transform(labeled_train[col_name]).toarray()

    col = ["bow_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    df_bow = pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)
    
    labeled_train = pd.concat([labeled_train, df_bow],axis=1)
    
    return labeled_train

## LDA

In [7]:
import gensim
from gensim import corpora, models
from gensim.corpora import TextCorpus, MmCorpus, Dictionary
from nltk.corpus import stopwords
from optparse import OptionParser

def make_lda(train, keep_n, num_topics, col_name):
    global used_model
    used_model = "lda"

    data = train[col_name]

    dictionary = corpora.Dictionary(data)
    dictionary.filter_extremes(keep_n=keep_n)

    corpus = [dictionary.doc2bow(text) for text in data]

    print("Make Lda..")

    lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=num_topics, chunksize=1000, passes=1)
    
    num = len(train)    
    df = []
    
    for i in range(0,num):
        if i % 10000 == 0:
            print(i)
        
        temp = [i[1] for i in lda.get_document_topics(corpus[i],minimum_probability=0)]
        df.append(temp)
    
    col = ["lda_%s_%d" % (col_name, data) for data in range(0, num_topics)]
    df = pd.DataFrame(df, columns = col)
    df.index = train.index
    
    train = pd.concat([train, df], axis=1)
        
    return train

## TF-IDF

In [8]:
from sklearn.feature_extraction.text import TfidfVectorizer

def make_tf_idf(train, max_features, col_name):
    global used_model
    used_model = "tf-idf"
    
    vectorizer = TfidfVectorizer(min_df = 1, max_features = max_features)
    
    train_data_features = vectorizer.fit_transform(train[col_name]).toarray()
    
    col = ["tfidf_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    df_tfidf = pd.DataFrame(train_data_features, columns = col, index=train.index)
    
    train = pd.concat([train, df_tfidf],axis=1)
    
    return train

In [9]:
max_features = 1000
#%time labeled_train = make_lda(labeled_train, 5000, 1000, "text_pos")
#%time labeled_train = make_bag_of_words(labeled_train, 1000, "text_pos_sentences")
labeled_train = make_tf_idf(labeled_train, max_features, "text_pos_sentences")

### predictor, model 생성

In [10]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['text_pos', 'text_pos_sentences', label])

model = RandomForestClassifier(n_estimators=10, n_jobs=-1)

### cross validation

In [11]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")
cv_result = scores.mean()

print(cv_result)

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.f

0.683697553571


/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.f

## Result
  * text의 morphs를 bag of words로 변환 (feature : 1,000개) - 0.637526911627
  * text의 pos를 bag of words로 변환 (feature : 1,000개) - 0.661265083065
  * 데이터를 셔플하고 text의 pos를 bag of words로 변환 (feature : 1,000개) - 0.688875771784
  
  * 데이터를 셔플하고 text의 pos를 tf-idf로 변환 (feature : 1,000개) - 0.683697553571
  
  * text의 pos를 lda로 변환 (keep_n : 5,000개, num_topics : 1,000개) - 0.643143549702
  * bs의 파라미터를 "lxml"이 아닌 "html.parser"로 주고 pos를 bag of words로 변환 (feature : 1,000개) - 0.657726939833

In [12]:
import os
import pickle

if not os.path.exists(used_model):
    os.makedirs(used_model)

pickle.dump(labeled_train[pre], open("%s/text_%d.p" % (used_model, max_features), "wb"), protocol = pickle.HIGHEST_PROTOCOL)